smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java

```diff
@@ -0,0 +1,182 @@
+        @Override
+        public void processStanza(Stanza packet) {
+            addOriginId(packet);
+        }
+    };
+
+    static {
+        XMPPConnectionRegistry.addConnectionCreationListener(new ConnectionCreationListener() {
+            @Override
+            public void connectionCreated(XMPPConnection connection) {
+                getInstanceFor(connection);
+            }
+        });
+    }
+
+    /**
+     * Private constructor.
+     * @param connection
+     */
+    private StableUniqueStanzaIdManager(XMPPConnection connection) {
+        super(connection);
+        enable();
+    }
+
+    /**
+     * Return an instance of the StableUniqueStanzaIdManager for the given connection.
+     *
+     * @param connection xmpp-connection
+     * @return manager instance for the connection
+     */
+    public static StableUniqueStanzaIdManager getInstanceFor(XMPPConnection connection) {
+        StableUniqueStanzaIdManager manager = INSTANCES.get(connection);
+        if (manager == null) {
+            manager = new StableUniqueStanzaIdManager(connection);
+            INSTANCES.put(connection, manager);
+        }
+        return manager;
+    }
+
+    /**
+     * Start appending origin-id elements to outgoing stanzas and add the feature to disco.
+     */
+    public void enable() {
+        ServiceDiscoveryManager disco = ServiceDiscoveryManager.getInstanceFor(connection());
+        if (disco.includesFeature(NAMESPACE)) {
+            return;
+        }
+        disco.addFeature(NAMESPACE);
+        connection().addPacketInterceptor(stanzaListener, OUTGOING_FILTER);
+    }
+
+    /**
+     * Stop appending origin-id elements to outgoing stanzas and remove the feature from disco.
+     */
+    public void disable() {
+        ServiceDiscoveryManager disco = ServiceDiscoveryManager.getInstanceFor(connection());
+        if (!disco.includesFeature(NAMESPACE)) {
+            return;
+        }
+        ServiceDiscoveryManager.getInstanceFor(connection()).removeFeature(NAMESPACE);
+        connection().removePacketInterceptor(stanzaListener);
+    }
+
+    /**
+     * Return true, if we automatically append origin-id elements to outgoing stanzas.
+     *
+     * @return true if functionality is enabled, otherwise false.
+     */
+    public boolean isEnabled() {
+        ServiceDiscoveryManager disco = ServiceDiscoveryManager.getInstanceFor(connection());
+        return disco.includesFeature(NAMESPACE);
+    }
+
+    /**
+     * Create an origin-id element with a random UUID.
+     *
+     * @return origin-id element.
+     */
+    public static OriginIdElement createOriginId() {
+        return new OriginIdElement(UUID.randomUUID().toString());
+    }
+
+    /**
+     * Add an origin-id element to a stanza and set the stanzas id to the same id as in the origin-id element.
+     *
+     * @param stanza stanza.
+     */
+    public static void addOriginId(Stanza stanza) {
+        OriginIdElement originId = createOriginId();
```

Comment:
Should be part of `OriginIdElement` API, not the manger. As a rule of thumb: Static methods in managers dealing soley with elements usually belong to those elements.